server/src/main/java/org/elasticsearch/index/mapper/vectors/NormalizedCosineFloatVectorValues.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper.vectors;
+
+import org.apache.lucene.index.FloatVectorValues;
+import org.apache.lucene.index.NumericDocValues;
+
+import java.io.IOException;
+
+public class NormalizedCosineFloatVectorValues extends FloatVectorValues {
+
+    private final FloatVectorValues in;
+    private final NumericDocValues magnitudeIn;
+    private final float[] vector;
+    private float magnitude = 1f;
+    private boolean hasMagnitude;
+    private int docId = -1;
+
+    public NormalizedCosineFloatVectorValues(FloatVectorValues in, NumericDocValues magnitudeIn) {
+        this.in = in;
+        this.magnitudeIn = magnitudeIn;
+        this.vector = new float[in.dimension()];
+    }
+
+    @Override
+    public int dimension() {
+        return in.dimension();
+    }
+
+    @Override
+    public int size() {
+        return in.size();
+    }
+
+    @Override
+    public float[] vectorValue() throws IOException {
+        // Lazy load vectors as we may iterate but not actually require the vector
+        return vectorValue(in.docID());
+    }
+
+    @Override
+    public int docID() {
+        return in.docID();
+    }
+
+    @Override
+    public int nextDoc() throws IOException {
+        int docId = in.nextDoc();
+        hasMagnitude = decodedMagnitude(docId);
+        return docId;
+    }
+
+    @Override
+    public int advance(int target) throws IOException {
+        int docId = in.advance(target);
+        hasMagnitude = decodedMagnitude(docId);
+        return docId;
+    }
+
+    public float magnitude() {
+        return magnitude;
+    }
+
+    private float[] vectorValue(int docId) throws IOException {
+        if (docId != this.docId) {
+            this.docId = docId;
+            // We should only copy and transform if we have a stored magnitude
+            if (hasMagnitude) {
+                System.arraycopy(in.vectorValue(), 0, vector, 0, dimension());
+                for (int i = 0; i < vector.length; i++) {
+                    vector[i] *= magnitude;
+                }
+                return vector;
+            } else {
+                return in.vectorValue();
+            }
+        } else {
+            return hasMagnitude ? vector : in.vectorValue();
+        }
+    }
+
+    private boolean decodedMagnitude(int docId) throws IOException {
+        if (magnitudeIn == null) {
+            return false;
+        }
+        int currentDoc = magnitudeIn.docID();
+        if (currentDoc == NO_MORE_DOCS || docId < currentDoc) {
```

Comment:
I don't think that `docId` can ever be less than `currentDoc`?

Also if we called `decodedMagnitude` in vectorValue() rather than nextDoc()/advance() then we wouldn't need to check if `currentDoc == NO_MORE_DOCS`?

Reply:
Good thinking! I was still seeing some slight discrepancy in my performance testings (nothing nearly like it was before) and I bet me attempting to decode and read a magnitude value on every doc instead of lazy loading it like the vector was causing the issue.